torch/onnx/utils.py

```diff
@@ -1004,35 +1019,98 @@ def _run_symbolic_function(g, block, n, inputs, env, operator_export_type=Operat
                     #
-                    # For `If` node, there is nothing to copy.
-                    # For `Loop` node, copy metadata for `iter`, `input_1`, ..., `input_n`.
-                    for i, b_in in enumerate(b.inputs()):
-                        if i == 0 and i < len(inputs):
-                            b_in.setType(inputs[i].type())
-                        if i > 0 and (i + 1) < len(inputs):
-                            b_in.setType(inputs[i + 1].type())
-                    torch._C._jit_pass_onnx_block(b, new_block, operator_export_type, env)
-                new_op_outputs = torch._C._jit_pass_fixup_onnx_controlflow_node(new_node, opset_version)
-                # Process Loop and If after subblock is converted.
-                from torch.onnx.symbolic_helper import _onnx_shape_inference
-                if _onnx_shape_inference:
-                    torch._C._jit_pass_onnx_node_shape_type_inference(new_node, _params_dict, opset_version)
-                return new_op_outputs
+                    # env: 65:10, 21:14, 22:10, input.53:16, weight:17
+                    # embedding_matrix.1 and input.1 are prim::Param, so we need convert to ONNX Identity to do shape inference.
+                    torch._C._jit_pass_onnx_node_shape_type_inference(inputs[0].node(), _params_dict, opset_version,
+                                                                      get_static_input_shape())
+                    static_if = True
+                    input_flag = inputs[0].node()['value'].tolist()
+                    if isinstance(input_flag, list):
+                        const_value = all(input_flag)
+                    else:
+                        const_value = bool(input_flag)
+                    block_idx = 0 if const_value else 1
+                    current_b = list(n.blocks())[block_idx]
+                    is_sub_block = True
+                    env = torch._C._jit_pass_onnx_block(current_b, block, operator_export_type, env, get_static_input_shape(),
+                                                        _params_dict, opset_version, enable_const_folding, is_sub_block)
+                    if_output_list = list(n.outputs())
+                    current_b_list = list(current_b.outputs())
+                    final_b_list = []
+                    for idx in range(len(if_output_list)):
+                        if current_b_list[idx] not in env:
+                            raise RuntimeError("The sub block ATen output " + current_b_list[idx] + " is not in env")
+                        if "onnx" in env[current_b_list[idx]].node().kind():
+                            onnx_b = env[current_b_list[idx]]
+                        else:
+                            onnx_b = g.op("Identity", env[current_b_list[idx]])
+                        final_b_list.append(onnx_b)
+                        torch._C._jit_pass_onnx_node_shape_type_inference(onnx_b.node(), _params_dict, opset_version,
```

Comment:
I see that you want to ensure shape inference was called, where output was not created by node inside this block. However, even if outside the block, the owner node should have shape inference called when being added to the graph right? The only exception might be graph input node (`prim::Param`), which we can probably do something special for it, like put its shape values into the map at initialization. Are there any other exceptions?

Reply:
Right, this shape inference is not needed.  Shape inference is called when the node gets added to the graph, so it is already done. Remove this one.